api/provider/aws/deploy_create.go

```diff
@@ -0,0 +1,139 @@
+package aws
+
+import (
+	"encoding/json"
+	"fmt"
+	"strconv"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/ecs"
+	"github.com/quintilesims/layer0/common/errors"
+	"github.com/quintilesims/layer0/common/models"
+)
+
+func (d *DeployProvider) Create(req models.CreateDeployRequest) (*models.Deploy, error) {
+	deployID := generateEntityID(req.DeployName)
+	familyName := addLayer0Prefix(d.Config.Instance(), req.DeployName)
+
+	taskDefinition, err := d.renderTaskDefinition(req.DeployFile, familyName)
+	if err != nil {
+		return nil, err
+	}
+
+	taskDefinitionOutput, err := d.createTaskDefinition(taskDefinition)
+	if err != nil {
+		return nil, err
+	}
+
+	bytes, err := json.Marshal(taskDefinitionOutput)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to extract deploy file: %s", err.Error())
+	}
+
+	deploy := &models.Deploy{
+		DeployName: req.DeployName,
+		DeployID:   deployID,
+		Version:    strconv.FormatInt(aws.Int64Value(taskDefinitionOutput.Revision), 10),
+		DeployFile: bytes,
+	}
+
+	taskDefinitionArn := aws.StringValue(taskDefinitionOutput.TaskDefinitionArn)
+
+	if err := d.createTags(deploy.DeployName, deploy.DeployID, deploy.Version, taskDefinitionArn); err != nil {
+		return deploy, err
+	}
+
+	return deploy, nil
+}
+
+func (d *DeployProvider) createTaskDefinition(taskDefinitionRequest *ecs.TaskDefinition) (*ecs.TaskDefinition, error) {
+
+	input := &ecs.RegisterTaskDefinitionInput{}
+	input.SetFamily(aws.StringValue(taskDefinitionRequest.Family))
+	input.SetTaskRoleArn(aws.StringValue(taskDefinitionRequest.TaskRoleArn))
+	input.SetNetworkMode(aws.StringValue(taskDefinitionRequest.NetworkMode))
+	input.SetContainerDefinitions(taskDefinitionRequest.ContainerDefinitions)
+	input.SetVolumes(taskDefinitionRequest.Volumes)
+	input.SetPlacementConstraints(taskDefinitionRequest.PlacementConstraints)
+
+	if err := input.Validate(); err != nil {
+		return nil, err
+	}
+
+	output, err := d.AWS.ECS.RegisterTaskDefinition(input)
+	if err != nil {
+		return nil, err
+	}
+
+	return output.TaskDefinition, nil
+}
+
+func (d *DeployProvider) renderTaskDefinition(body []byte, familyName string) (*ecs.TaskDefinition, error) {
+	var taskDefinition *ecs.TaskDefinition
+
+	if err := json.Unmarshal(body, &taskDefinition); err != nil {
+		return nil, fmt.Errorf("Failed to decode deploy: %s", err.Error())
+	}
+
+	taskDefinition.SetFamily(familyName)
+
+	if len(taskDefinition.ContainerDefinitions) == 0 {
+		return nil, fmt.Errorf("Deploy must have at least one container definition")
+	}
+
+	if taskDefinition.Family != nil && aws.StringValue(taskDefinition.Family) != familyName {
+		return nil, errors.Newf(errors.InvalidRequest, "Custom family names are unsupported in Layer0")
+	}
+
+	// Valid NetworkMode options are [bridge, host, none], "bridge" is most common so defaulting to that.
+	if taskDefinition.NetworkMode == nil {
```

Comment:
do we need this? Can't we let aws handle the default? Seems to be fine in the past. What if users don't want networking mode options?

Reply:
In testing, AWS was requiring that `NetworkMode` was in the task definition, otherwise the error "NetworkMode must be specified" was returned. If a user wants no networking mode, they can select `none`.